apiv2/db_import/importers/base_importer.py

```diff
@@ -0,0 +1,299 @@
+        pass
+
+    def import_to_db(self) -> Base:
+        """
+        Gets the mapping, and queries to check if the table already has a record matching the id fields. If not, it will
+        create a new object and insert it, else it will update the object with values from the metadata.
+        """
+        logger.debug(json.dumps(self.metadata, indent=2))
+        data_map = self.get_data_map()
+        identifiers = {id_field: map_to_value(id_field, data_map, self.metadata) for id_field in self.get_id_fields()}
+        klass = self.get_db_model_class()
+        session = self.config.get_db_session()
+        try:
+            query = sa.select(klass).where(sa.and_(*[getattr(klass, k) == v for k, v in identifiers.items()]))
+            db_obj = session.scalars(query).one()
+        except NoResultFound:
+            db_obj = klass()
+
+        for db_key, _data_path in data_map.items():
+            setattr(db_obj, db_key, map_to_value(db_key, data_map, self.metadata))
+
+        session.add(db_obj)
+        session.flush()
+        return db_obj
+
+
+class StaleDeletionDBImporter(BaseDBImporter):
+    """
+    Supports insert of new record, update of existing record, and delete of stale records. This class iterates over a
+    list of individual records. This is a good fit for tables that aren't referenced as foreign keys.
+    """
+
+    def get_filters(self) -> dict[str, str]:
+        """
+        Provides filter criteria for fetching all the sibling records.
+        Ex: filter by dataset_id=XYZ on dataset_authors to fetch all authors of dataset XYZ.
+        """
+        pass
+
+    @classmethod
+    def get_hash_value(cls, record: Base) -> str:
+        """
+        Generates hash value for a record from values of its id fields separated by "-".
+        """
+        return "-".join([f"{getattr(record, attr)}" for attr in cls.get_id_fields()])
+
+    def get_existing_objects(self) -> tuple[list[Base], dict[str, Base]]:
+        """
+        Creates a map of existing items by querying using the filter criteria. The map is keyed on hash value of the
+        record.
+        """
+        hashed_records = {}
+        klass = self.get_db_model_class()
+        session = self.config.get_db_session()
+        query = session.scalars(
+            sa.select(klass).where(sa.and_(*[getattr(klass, k) == v for k, v in self.get_filters().items()])),
+        ).all()
+        all_records = []
+        for record in query:
+            all_records.append(record)
+            hashed_records[self.get_hash_value(record)] = record
+        return all_records, hashed_records
+
+    def update_data_map(self, data_map: dict[str, Any], metadata: dict[str, Any], index: int) -> dict[str, Any]:
+        """
+        Method to be overridden in cases where custom data mapping is necessary based on index value of the record in
+        the list.
+        """
+        return data_map
+
+    def import_to_db(self) -> Base:
```

Comment:
Most of the code copied over from the v1 schema is just minor tweaks to reflect schema and sqlalchemy changes, but I made pretty significant changes to this `StaleDeletionDBImporter` class since it wasn't handling cases where there were multiple rows for a hash key. This could use 👀 and comparison to the old one.